libvast/vast/flatbuffer.hpp

```diff
@@ -0,0 +1,350 @@
+  // *file_identifier* from *Table*, which makes this API a lot easier to use.
+  // When upgrading to that version from 1.12 we can enable reflection when
+  // compiling the schemas.
+  flatbuffer(flatbuffers::FlatBufferBuilder& builder,
+             flatbuffers::Offset<Table> offset,
+             const char* file_identifier) noexcept requires(IsRootTable) {
+    builder.Finish(offset, file_identifier);
+    auto chunk = chunk::make(builder.Release());
+    VAST_ASSERT(chunk);
+    *this = flatbuffer{chunk};
+  }
+
+  /// Constructs a ref-counted FlatBuffers table that shares the
+  /// lifetime with another FlatBuffer pointer.
+  /// @pre `parent`
+  /// @pre *table* must be accessible from *parent*.
+  template <class ParentTable, bool ParentIsRootTable>
+  flatbuffer(flatbuffer<ParentTable, ParentIsRootTable> parent,
+             const Table& table) noexcept requires(!IsRootTable)
+    : chunk_{std::exchange(parent.chunk_, {})}, table_{&table} {
+    VAST_ASSERT(chunk_);
+    VAST_ASSERT(reinterpret_cast<const std::byte*>(table_) >= chunk_->data());
+    VAST_ASSERT(reinterpret_cast<const std::byte*>(table_)
+                < (chunk_->data() + chunk_->size()));
+  }
+
+  ~flatbuffer() noexcept = default;
+
+  flatbuffer(const flatbuffer& other) noexcept
+    : chunk_{other.chunk_}, table_{other.table_} {
+    // nop
+  }
+
+  flatbuffer(flatbuffer&& other) noexcept
+    : chunk_{std::exchange(other.chunk_, {})},
+      table_{std::exchange(other.table_, {})} {
+    // nop
+  }
+
+  flatbuffer(std::nullptr_t) noexcept {
+    // nop
+  }
+
+  flatbuffer& operator=(const flatbuffer& rhs) noexcept {
+    if (&rhs == this)
+      return *this;
+    chunk_ = rhs.chunk_;
+    table_ = rhs.table_;
+    return *this;
+  }
+
+  flatbuffer& operator=(flatbuffer&& rhs) noexcept {
+    chunk_ = std::exchange(rhs.chunk_, {});
+    table_ = std::exchange(rhs.table_, {});
+    return *this;
+  }
+
+  flatbuffer& operator=(std::nullptr_t) noexcept {
+    chunk_ = nullptr;
+    table_ = nullptr;
+    return *this;
+  }
+
+  // -- operators -------------------------------------------------------------
+
+  explicit operator bool() const noexcept {
+    return table_ != nullptr;
+  }
+
+  const Table& operator*() const noexcept {
+    VAST_ASSERT(table_);
+    return *table_;
+  }
+
+  const Table* operator->() const noexcept {
+    VAST_ASSERT(table_);
+    return table_;
+  }
+
+  friend bool operator==(flatbuffer lhs, flatbuffer rhs) noexcept
+    requires(IsRootTable) {
+    if (&lhs == &rhs)
+      return true;
+    if (lhs && rhs) {
+      const auto lhs_bytes = as_bytes(lhs);
+      const auto rhs_bytes = as_bytes(rhs);
+      return std::equal(lhs_bytes.begin(), lhs_bytes.end(), rhs_bytes.begin(),
+                        rhs_bytes.end());
+    }
+    return static_cast<bool>(lhs) == static_cast<bool>(rhs);
+  }
+
+  friend std::strong_ordering
+  operator<=>(flatbuffer lhs, flatbuffer rhs) noexcept requires(IsRootTable) {
```

Comment:
I'm not sure I understand the semantics of a flatbuffer being "bigger" or "smaller" than another? Is this required somewhere? Also, I think this will lead to counter-intuitive results if you have e.g.

```
table Integer {
  value: uint32;
}
```

since the ordering would not follow the numerical ordering of value.

Reply:
We can maybe make this an opt-in thing via a third template argument that toggles the availability of "behaves like a `vector<byte>` for comparisons."

Reply:
>  Is this required somewhere

The new type system does the same thing internally because the types need to be sortable to be used in an unordered container. This makes it easy to use that way. I think we can make it an option, though.